File: Quiz1/Q1.cpp

#include <string>
#include <iostream>
#include <vector>

using namespace std;
// 400101204
// M. parsa Dini
bool my_bfs(int i, int j, int index, vector<vector<char>>& grid, string& name) {
    if (index == name.length()) { // naem getting found
        return true;
    }
    if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() || grid[i][j] != name[index]) {
        return false; //outside of the bloody grid
    }
    char temp = grid[i][j]; //current char
    grid[i][j] = '#';
    bool found = my_bfs(i + 1, j, index + 1, grid, name) ||//R
                 my_bfs(i - 1, j, index + 1, grid, name) || //L
                 my_bfs(i, j + 1, index + 1, grid, name) ||//U
                 my_bfs(i, j - 1, index + 1, grid, name); //D
    grid[i][j] = temp; // my last char
    return found;
}

int main() {
    //handling inputs
    int n, m;
    string name;
    cin >> n >> m;
    cin >> name;

    // worst criteria
    if (name.length() > n * m) {
        cout << "nabood!" << endl;
        return 0;
    }

    vector grid(n, vector<char>(m));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> grid[i][j];
        }
    }
    // the alg
    bool name_found = false;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (grid[i][j] == name[0]) {
                if (my_bfs(i, j, 0, grid, name)){
                    name_found = true;
                    break;
                }
            }
        }
        if (name_found) {
            break;
        }
    }

    if (name_found) {
        cout << "peyda kardam!" << endl;
    } else {
        cout << "nabood!" << endl;
    }

    return 0;
}

File: Quiz1/Q1.java
import java.util.Scanner;

public class Main {

    public static boolean myBFS(int i, int j, int index, char[][] grid, String name) {
        if (index == name.length()) {
            return true; // Name fully matched
        }
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || grid[i][j] != name.charAt(index)) {
            return false; // Outside grid or character mismatch
        }

        char temp = grid[i][j]; // Current character
        grid[i][j] = '#'; // Mark visited

        boolean found = myBFS(i + 1, j, index + 1, grid, name) || // Down
                        myBFS(i - 1, j, index + 1, grid, name) || // Up
                        myBFS(i, j + 1, index + 1, grid, name) || // Right
                        myBFS(i, j - 1, index + 1, grid, name);  // Left

        grid[i][j] = temp; // Reset cell
        return found;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Handle input
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        scanner.nextLine(); // Consume the newline
        String name = scanner.nextLine();

        // Quick rejection if name length exceeds grid size
        if (name.length() > n * m) {
            System.out.println("nabood!");
            scanner.close();
            return;
        }

        char[][] grid = new char[n][m];
        for (int i = 0; i < n; i++) {
            String line = scanner.nextLine();
            for (int j = 0; j < m; j++) {
                grid[i][j] = line.charAt(j);
            }
        }

        // Algorithm
        boolean nameFound = false;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == name.charAt(0)) {
                    if (myBFS(i, j, 0, grid, name)) {
                        nameFound = true;
                        break;
                    }
                }
            }
            if (nameFound) {
                break;
            }
        }

        if (nameFound) {
            System.out.println("peyda kardam!");
        } else {
            System.out.println("nabood!");
        }

        scanner.close();
    }
}
